File: web/src/api/badges.ts
import client from "./client";

export interface Badge {
  id: number;
  name: string;
  description: string;
  icon: string;
  earned: boolean;
  awarded_at: string | null;
}

export const listBadges = () =>
  client.get<Badge[]>("/badges").then((r) => r.data);

export const evaluateBadges = () =>
  client.post<{ newly_awarded: string[] }>("/badges/evaluate").then((r) => r.data);

export const seedBadges = () =>
  client.post("/badges/seed").then((r) => r.data);

File: apps/app/schemas/badges.py
from datetime import datetime

from pydantic import BaseModel


class BadgeDefinitionResponse(BaseModel):
    id: int
    name: str
    description: str
    icon: str

    model_config = {"from_attributes": True}


class UserBadgeResponse(BaseModel):
    id: int
    badge: BadgeDefinitionResponse
    awarded_at: datetime

    model_config = {"from_attributes": True}
